### PR TITLE
custom authority，schema can be assigned to proxy users

### PR DESCRIPTION
**Pull request by lwclover**
authority enhance，sample：
```
- !AUTHORITY
  users:
    - root@:root
    - user1@:user1
    - user1@127.0.0.1:user1
  provider:
    type: SCHEMA_PRIVILEGES_PERMITTED
    props:
      user-schema-mappings: root@=test, user1@127.0.0.1=db_dal_admin, user1@=test, user1@=db_dal_admin
```

when using root login from any host，you can use database test.
when using user1 login from 127.0.0.1，you can use database db_dal_admin.
when using user1 login from any other host，you can use database test and db_dal_admin.

### CODE REVIEW

**Comment by tristaZero** [comment]
Hi @lwclover ,

Thanks for your PR, which attracts my interest. Let firstly set aside your coding, and target your idea and plan.

It basically looks nice to me, but there is an oversight, i.e. when users change the map of user and privileges, `refresh()` just refresh the runtime. All these changes will lose once users restart ShardingProxy. What do you think?

**Review by tristaZero — CHANGES_REQUESTED** [review]
Hi @lwclover ,

Thanks for your PR, which attracts my interest. Let set aside your coding, and firstly target your idea and plan.

It basically looks nice, but there is an oversight, i.e. when users change the map of user and privileges, refresh() just refresh the runtime. All these changes will lose once users restart ShardingProxy. What do you think?

**Comment by lwclover** [comment]
> Hi @lwclover ,
> 
> Thanks for your PR, which attracts my interest. Let set aside your coding, and firstly target your idea and plan.
> 
> It basically looks nice, but there is an oversight, i.e. when users change the map of user and privileges, refresh() just refresh the runtime. All these changes will lose once users restart ShardingProxy. What do you think?

user-schema-mappings persists on registry center like other configs.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/ShowDatabasesExecutor.java [hunk 1]
```diff
@@ -54,18 +54,18 @@ public void execute(final BackendConnection backendConnection) {
...
+            SQLCheckEngine.check(new MySQLShowDatabasesStatement(), Collections.emptyList(), contexts.getGlobalRuleMetaData().getRules(), backendConnection.getSchemaName(), contexts.getMetaDataMap(),
+                    backendConnection.getGrantee());
         } catch (final SQLCheckException ex) {
-            Collection<Object> result = new LinkedList<>();
-            for (String each : ProxyContext.getInstance().getAllSchemaNames()) {
-                if (SQLCheckEngine.check(each, getRules(each), backendConnection.getGrantee())) {
-                    result.add(each);
-                }
+            // ignore
+        }
```
Comment (tristaZero) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/ShowDatabasesExecutor.java:61]: Is there a need to do such modification?

File: shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/model/privilege/SchemaPrivilegesPermittedShardingSpherePrivileges.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Override
+    public boolean hasPrivileges(final AccessSubject accessSubject, final Collection<PrivilegeType> privileges) {
+        if (accessSubject instanceof SchemaAccessSubject) {
+            return hasPrivileges(((SchemaAccessSubject) accessSubject).getSchema());
+        }
+        throw new UnsupportedOperationException(accessSubject.getClass().getCanonicalName());
+    }
+
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/model/privilege/SchemaPrivilegesPermittedShardingSpherePrivileges.java:57]: Could you remove this redundant line?

File: shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/model/privilege/SchemaPrivilegesPermittedShardingSpherePrivileges.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import org.apache.shardingsphere.authority.model.AccessSubject;
+import org.apache.shardingsphere.authority.model.PrivilegeType;
+import org.apache.shardingsphere.authority.model.ShardingSpherePrivileges;
+import org.apache.shardingsphere.authority.provider.natived.model.subject.SchemaAccessSubject;
+
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public class SchemaPrivilegesPermittedShardingSpherePrivileges implements ShardingSpherePrivileges {
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/model/privilege/SchemaPrivilegesPermittedShardingSpherePrivileges.java:31]: Hi suppose it has no possibility to have chid class, please define it as a final class.

File: shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/builder/SchemaPrivilegeBuilder.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+     * Build privileges.
+     *
+     * @param users users
+     * @param props props
+     * @return privileges
+     */
+    public static Map<ShardingSphereUser, ShardingSpherePrivileges> build(final Collection<ShardingSphereUser> users, final Properties props) {
+        String mappingProp = props.getProperty(SchemaPrivilegesPermittedAuthorityProviderAlgorithm.PROP_USER_SCHEMA_MAPPINGS, "");
+        Preconditions.checkArgument(!"".equals(mappingProp), "user-schema-mappings configuration `%s` can not be null", mappingProp);
+
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/builder/SchemaPrivilegeBuilder.java:52]: Could you remove this redundant line?

File: shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/builder/SchemaPrivilegeBuilder.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    public static Map<ShardingSphereUser, ShardingSpherePrivileges> build(final Collection<ShardingSphereUser> users, final Properties props) {
+        String mappingProp = props.getProperty(SchemaPrivilegesPermittedAuthorityProviderAlgorithm.PROP_USER_SCHEMA_MAPPINGS, "");
+        Preconditions.checkArgument(!"".equals(mappingProp), "user-schema-mappings configuration `%s` can not be null", mappingProp);
+        
+        String[] mappings = mappingProp.split(",");
+        Arrays.asList(mappings).stream().forEach(each -> Preconditions.checkArgument(0 < each.indexOf("@") && 0 < each.indexOf("="), 
+                "user-schema-mappings configuration `%s` is invalid, the configuration format should be like `username@hostname=schema`", each));
+
+        Map<ShardingSphereUser, ShardingSpherePrivileges> result = new HashMap<>();
+
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/builder/SchemaPrivilegeBuilder.java:58]: Could you remove this redundant line?

File: shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/builder/SchemaPrivilegeBuilder.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+     * @return privileges
+     */
+    public static Map<ShardingSphereUser, ShardingSpherePrivileges> build(final Collection<ShardingSphereUser> users, final Properties props) {
+        String mappingProp = props.getProperty(SchemaPrivilegesPermittedAuthorityProviderAlgorithm.PROP_USER_SCHEMA_MAPPINGS, "");
+        Preconditions.checkArgument(!"".equals(mappingProp), "user-schema-mappings configuration `%s` can not be null", mappingProp);
+        
+        String[] mappings = mappingProp.split(",");
+        Arrays.asList(mappings).stream().forEach(each -> Preconditions.checkArgument(0 < each.indexOf("@") && 0 < each.indexOf("="), 
+                "user-schema-mappings configuration `%s` is invalid, the configuration format should be like `username@hostname=schema`", each));
+
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/builder/SchemaPrivilegeBuilder.java:56]: Could you remove this redundant line?

File: shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/builder/SchemaPrivilegeBuilder.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import org.apache.shardingsphere.authority.provider.schema.model.privilege.SchemaPrivilegesPermittedShardingSpherePrivileges;
+import org.apache.shardingsphere.infra.metadata.user.ShardingSphereUser;
+
+import com.google.common.base.Preconditions;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class SchemaPrivilegeBuilder {
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/builder/SchemaPrivilegeBuilder.java:41]: Hi it is suggested to define it as a final one.

File: shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/SchemaPrivilegesPermittedAuthorityProviderAlgorithm.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    @Override
+    public Optional<ShardingSpherePrivileges> findPrivileges(final Grantee grantee) {
+        return userPrivilegeMap.keySet().stream().filter(each -> each.getGrantee().equals(grantee)).findFirst().map(userPrivilegeMap::get);
+    }
+
+    @Override
+    public String getType() {
+        return "SCHEMA_PRIVILEGES_PERMITTED";
+    }
+
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-authority/shardingsphere-infra-authority-common/src/main/java/org/apache/shardingsphere/authority/provider/schema/SchemaPrivilegesPermittedAuthorityProviderAlgorithm.java:70]: Could you remove this redundant line?

**Review by tristaZero — CHANGES_REQUESTED** [review]
Hey @lwclover Thanks for this enhanced implementation. I suppose it will be ready to merge after these comments are fixed. Thanks.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/ShowDatabasesExecutor.java (hunk 1 above)
Comment (tristaZero) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/ShowDatabasesExecutor.java:61]: Hey @terrymanu 
@lwclover and I were confused here, actually it is not straightforward enough, could you have a look?

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/ShowDatabasesExecutor.java (hunk 1 above)
Comment (terrymanu) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/ShowDatabasesExecutor.java:61]: Maybe cause problem if SQLCheckException occur, why ignore the exception here?

Comment (lwclover) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/admin/mysql/executor/ShowDatabasesExecutor.java:61]: I think codes below are redundant lines. 
```
MetaDataContexts contexts = ProxyContext.getInstance().getMetaDataContexts();
            SQLCheckEngine.check(new MySQLShowDatabasesStatement(), Collections.emptyList(), contexts.getGlobalRuleMetaData().getRules(), backendConnection.getSchemaName(), contexts.getMetaDataMap(),
                    backendConnection.getGrantee());
```

**Review by terrymanu — APPROVED** [review]

**Review by terrymanu — CHANGES_REQUESTED** [review]
Please submit pr into master branch

**Comment by lwclover** [comment]
changes submit to master

**Pull request closed without merging**